lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp

```diff
@@ -0,0 +1,348 @@
+        SubstFlags::SubstituteOpaqueArchetypes | SubstFlags::AllowLoweredTypes);
+  }
+};
+} // namespace
+
+void OpaqueSpecializerCloner::clone() {
+  for (auto arg: entryBlock->getArguments())
+    recordFoldedValue(arg, arg);
+  cloneReachableBlocks(cloneFromBlock, {}, entryBlock,
+                       true /*havePrepopulatedFunctionArgs*/);
+  getBuilder().setInsertionPoint(entryBlock);
+  getBuilder().createBranch(RegularLocation::getAutoGeneratedLocation(),
+                            getOpBasicBlock(cloneFromBlock));
+}
+
+/// Update address uses of the opaque type archetype with the concrete type.
+/// This is neccessary for apply instructions.
+void OpaqueSpecializerCloner::insertOpaqueToConcreteAddressCasts(
+    SILInstruction *orig, SILInstruction *cloned) {
+
+  // Replace apply operands.
+  if (auto apply = ApplySite::isa(cloned)) {
+    SavedInsertionPointRAII restore(getBuilder());
+    getBuilder().setInsertionPoint(apply.getInstruction());
+    auto substConv = apply.getSubstCalleeConv();
+    unsigned idx = 0;
+    for (auto &opd : apply.getArgumentOperands()) {
+      auto argConv = apply.getArgumentConvention(opd);
+      auto argIdx = apply.getCalleeArgIndex(opd);
+      auto argType = substConv.getSILArgumentType(argIdx);
+      if (argType.getASTType() != opd.get()->getType().getASTType()) {
+        if (argConv.isIndirectConvention()) {
+          auto cast = getBuilder().createUncheckedAddrCast(apply.getLoc(),
+                                                           opd.get(), argType);
+          opd.set(cast);
+        } else {
+          auto cast = getBuilder().createUncheckedRefCast(apply.getLoc(),
+                                                          opd.get(), argType);
+          opd.set(cast);
+        }
+      }
+      ++idx;
+    }
+    return;
+  }
+}
+
+namespace {
+class OpaqueArchetypeSpecializer : public SILFunctionTransform {
+  void run() override {
+    auto *currentModule = getFunction()->getModule().getSwiftModule();
+
+    auto opaqueArchetypeWouldChange = [=](CanType ty) -> bool {
+      if (!ty->hasOpaqueArchetype())
+        return false;
+      ReplaceOpaqueTypesWithUnderlyingTypes replacer(currentModule);
+      return ty.subst(replacer, replacer,
+                      SubstFlags::SubstituteOpaqueArchetypes |
+                          SubstFlags::AllowLoweredTypes)
+                 ->getCanonicalType() != ty;
+    };
+
+    // Look for opaque type archetypes.
+    bool foundOpaqueArchetype = false;
+    for (auto &BB : *getFunction()) {
+      for (auto &inst : BB) {
+        for (auto &opd : inst.getAllOperands()) {
+          if (!opaqueArchetypeWouldChange(opd.get()->getType().getASTType()))
+            continue;
+          foundOpaqueArchetype = true;
+          break;
+        }
+        if (foundOpaqueArchetype)
+          break;
+        auto *allocStack = dyn_cast<AllocStackInst>(&inst);
+        if (!allocStack || !opaqueArchetypeWouldChange(
```

Comment:
I don't understand why AllocStackInsts are special. You should specialize a function even if it doesn't contain any AllocStackInsts